### DISCUSSION TITLE
Pydantic Validation without Error Raising

### DISCUSSION THREAD

**Discussion by joe-sciame-wm** [pydantic/pydantic#5261] [Question]
Hi Pydantic Team,

Is there a way in Pydantic to have validations that allow custom logger and not throw an error?  I'm working on some Pydantic classes for existing data and I want to log all the current missing and incorrectly formatted values with the existing data and start capturing the issues and then eventually move over to raising errors for the validation errors.

The Pydantic model is fairly complex with a top level class and many nested sub classes.  What I want to do is get a return of all the validations errors that would have occurred from all the subclass validations, but still get the model back so I can work with it. I'm making the model flexible enough so that Pydantic can parse the worst case, but then adding validations on top to start logging the errors in the model with the goal of fixing them overtime.

**Comment by dmontagu** [comment]
How performance sensitive are you?

One possible idea: copy the definitions of the classes but change them in such a way that it will pass validation for all the data you expect to receive, by changing the attributes to be of type Any or similar.

Then, in a pre=True root_validator on the parent, attempt to validate the data against the "strict" version of the class. Just do a `try: except ValidationError:` and log out the errors however you like, and return the input data unmodified. If that's not clear enough, let me know.

While defining strict/lenient versions of the class may be an annoying amount of work, I think there's a _ton_ of edge cases to consider/behaviors to bikeshed over if we were to try to implement a way to allow certain kinds of validation errors through with handling besides raising an error. I'm not opposed in principle if someone figures out a nice way to allow it that isn't a big change to pydantic, but my intuition is that it would be hard to implement/support.

**Reply by joe-sciame-wm** [reply]
Thanks! What I've tried so far is to create a private attribute to hold the validation errors. 

```
class BaseModel(PydanticBaseModel):
    _validation_errors: Optional[util.ValidationErrors]

    class Config:
        alias_generator = to_camel
        allow_population_by_field_name = True
        underscore_attrs_are_private = True
        arbitrary_types_allowed = True

    @classmethod
    def _process_error(cls, error: exceptions.HDRMetadataValidationError,
                       errors: util.ValidationErrors = util.ValidationErrors()) -> util.ValidationErrors:
        errors.add_error(error)
        logging.warning(error)
        return errors
```
Then in my validation code, I'm trying to add the private attribute in the data and have it be brought into the class. It doesn't work for me as it doesn't look like the private attribute is parsed from the dictionary like the other classes.  

```
    @root_validator(pre=True)
    def validate_hdr_metadata(cls, values):
        errors = util.ValidationErrors()
        if not values.get('hdrMetadata'):
            errors = cls._process_error(
                exceptions.HDRMetadataValidationError('HDR Metadata Missing'),
                errors=errors)
        if values.get('hdrMetadata') and (
                int(values['hdrMetadata']['MaxFALL']) and int(
            values['hdrMetadata']['MaxCLL'])):
            errors = cls._process_error(exceptions.HDRMetadataValidationError(
                'CPL Metadata exists but is 0, 0'), errors=errors)
        values['_validationErrors'] = errors
        return values
```

When I try to access the attribute, I see an exception

```
'Traceback (most recent call last):
  File "/Applications/PyCharm.app/Contents/plugins/python/helpers/pydev/_pydevd_bundle/pydevd_resolver.py", line 178, in _getPyDictionary
    attr = getattr(var, n)
AttributeError: _validation_errors
```
'

**Reply by joe-sciame-wm** [reply]
A solution that works is creating a validation_errors main attribute and then setting it to be excluded on dump. 

`class BaseModel(PydanticBaseModel):
    validation_errors: Optional[util.ValidationErrors] = Field(exclude=True)

    class Config:
        alias_generator = to_camel
        allow_population_by_field_name = True

    @classmethod
    def _process_error(cls, error: exceptions.HDRMetadataValidationError,
                       errors: util.ValidationErrors = util.ValidationErrors()) -> util.ValidationErrors:
        errors.add_error(error)
        logging.warning(error)
        return errors`

**Comment by sezanzeb** [comment]
I wanted to propose this:

```py
from pydantic import BaseModel, ValidationError

class B(BaseModel):
    qux: int

class A(BaseModel):
    foo: int
    bar: B

values = {
    "foo": 1,
    "bar": {
        "qux": "b"
    },
}

a = A.construct(**values)
try:
    A.validate(values)
except ValidationError as e:
    print(f"Error: Validation failed: {str(e)}")

print(a)

```

```
Error: Validation failed: 1 validation error for A
bar -> qux
  value is not a valid integer (type=type_error.integer)
foo=1 bar={'qux': 'b'}
```

But as you can see, `bar` is a dict when using `construct`. So this approach doesn't quite work for nested models. Tested in with 1.7.4 and 2.9.1. But for flat models this seems to be a possible solution, if doing the try/except like this is an option.

**Reply by sezanzeb** [reply]
the "dacite" library might be an alternative in this case. It can disable validation in its from_dict method: https://github.com/konradhalas/dacite/?tab=readme-ov-file#types-checking
